src/Pos/Wallet/Web/Tracking.hs

```diff
@@ -0,0 +1,280 @@
+import           Pos.Util.Chrono            (getNewestFirst)
+import qualified Pos.Util.Modifier          as MM
+
+import           Pos.Wallet.Web.ClientTypes (CAccountAddress (..), CAddress, WS,
+                                             addressToCAddress, encToCAddress)
+import           Pos.Wallet.Web.State       (WebWalletModeDB)
+import qualified Pos.Wallet.Web.State       as WS
+
+type CAccModifier = MM.MapModifier CAccountAddress ()
+
+type BlockLockMode ssc m =
+    ( SscHelpersClass ssc
+    , WithLogger m
+    , WithNodeContext ssc m
+    , MonadDB m
+    , MonadMask m)
+
+
+-- To support actual wallet accounts we listen to applications and rollbacks
+-- of blocks, extract transactions from block and extract our
+-- accounts (such accounts which we can decrypt).
+-- We synchronise wallet-db (acidic-state) with node-db
+-- and support last seen tip for each walletset.
+-- There are severals cases when we must  synchronise wallet-db and node-db:
+-- • When we relaunch wallet. Desynchronization can be caused by interruption
+--   during blocks application/rollback at the previous launch,
+--   then wallet-db can fall behind from node-db (when interruption during rollback)
+--   or vice versa (when interruption during application)
+--   @syncWSetsWithGStateLock@ implements this functionality.
+-- • When a user wants to import a secret key. Then we must rely on
+--   Utxo (GStateDB), because blockchain can be large.
+
+-- Select our accounts from Utxo and put to wallet-db.
+-- Used for importing of a secret key.
+selectAccountsFromUtxoLock
+    :: forall ssc m . (WebWalletModeDB m, BlockLockMode ssc m)
+    => [EncryptedSecretKey]
+    -> m ()
+selectAccountsFromUtxoLock encSKs = withBlkSemaphore $ \tip -> do
+    let (hdPass, wsAddr) = unzip $ map getEncInfo encSKs
+    addresses <- discoverHDAddresses hdPass
+    let allAddreses = concatMap createAccounts $ zip wsAddr addresses
+    tip <$ mapM_ WS.addAccount allAddreses
+  where
+    createAccounts :: (CAddress WS, [(Address, [Word32])]) -> [CAccountAddress]
+    createAccounts (wsAddr, addresses) = do
+        let (ads, paths) = unzip addresses
+        mapMaybe createAccount $ zip3 (repeat wsAddr) ads paths
+
+    createAccount :: (CAddress WS, Address, [Word32]) -> Maybe CAccountAddress
+    createAccount (wsAddr, addr, derPath) = do
+        guard $ length derPath == 2
+        pure $ CAccountAddress wsAddr (derPath !! 0) (derPath !! 1) (addressToCAddress addr)
+
+-- Iterate over blocks (using forward links) and actualize our accounts.
+syncWSetsWithGStateLock
+    :: forall ssc m . (WebWalletModeDB m, BlockLockMode ssc m)
+    => [EncryptedSecretKey]
+    -> m ()
+syncWSetsWithGStateLock encSKs = withBlkSemaphore $ \tip ->
+    tip <$ mapM_ (syncWSetsWithGState @ssc) encSKs
+
+-- | Run action acquiring lock on block application.
+-- Argument of action is an old tip, result is put as a new tip.
+-- Wallet version (without extra constraints)
```

Comment:
What's that? Two same functions? Not acceptable. Remove unnecessary constraints from original `withBlkSemaphore`.